### PR TITLE
Use custom attributes

### PR DESCRIPTION
**Pull request by tlein**
blocked by #212

### CODE REVIEW

File: src/duckling/controls/json.component.ts [hunk 1]
```diff
@@ -102,47 +129,35 @@ export class JsonComponent {
...
-    get jsonKeys() : string[] {
-        let keys : string[] = [];
-        for (let key in this.value) {
-            keys.push(key);
-        }
-        return keys;
+    
+    onArrayAddClicked(key : string) {
+        let patchArray = [];
+        patchArray[this.value[key].length] = getDefaultForSchema((this.schema[key] as JsonSchema[])[0] as JsonSchema);
```
Comment (JeffSwenson) [inline src/duckling/controls/json.component.ts]: D:\Jeff\Code\ILikeDucks\duckling\node_modules\@angular\core\bundles\core.umd.js:2843 TypeError: Cannot read property 'keys' of undefined
    at JsonSchemaEditComponent.getKeyIndices (D:\Jeff\Code\ILikeDucks\duckling\build\duckling\controls\json-schema-edit.component.js:164)
    at DebugAppView.View_JsonSchemaEditComponent0.detectChangesInternal (component.ngfactory.js:84)
    at DebugAppView.AppView.detectChanges (D:\Jeff\Code\ILikeDucks\duckling\node_modules\@angular\core\bundles\core.umd.js:9354)
    at DebugAppView.detectChanges (D:\Jeff\Code\ILikeDucks\duckling\node_modules\@angular\core\bundles\core.umd.js:9447)
    at DebugAppView.View_CustomAttributesComponent0.detectChangesInternal (component.ngfactory.js:274)
    at DebugAppView.AppView.detectChanges (D:\Jeff\Code\ILikeDucks\duckling\node_modules\@angular\core\bundles\core.umd.js:9354)
    at DebugAppView.detectChanges (D:\Jeff\Code\ILikeDucks\duckling\node_modules\@angular\core\bundles\core.umd.js:9447)
    at DebugAppView.View_CustomAttributesComponent_Host0.detectChangesInternal (host.ngfactory.js:29)
    at DebugAppView.AppView.detectChanges (D:\Jeff\Code\ILikeDucks\duckling\node_modules\@angular\core\bundles\core.umd.js:9354)
    at DebugAppView.detectChanges (D:\Jeff\Code\ILikeDucks\duckling\node_modules\@angular\core\bundles\core.umd.js:9447)

File: src/duckling/project/custom-attributes.component.ts [hunk 2]
```diff
@@ -83,26 +84,23 @@ export class CustomAttributesComponent {
...
         this.curAddingSchema = newJsonSchema;
     }
     
     onAcceptCustomAttribute() {
-        this._project.addCustomAttribute(this.curAddingName, schemaToJson(this.curAddingSchema));
-        this.curAddingName = "";
-        this.curAddingSchema = {keys: [], contents: []};
+        this._project.addCustomAttribute(this.curAddingName, schemaEditToJson(this.curAddingSchema));
+        this.curAddingName = undefined;
+        this.curAddingSchema = undefined;
```
Comment (jeffswenson) [inline src/duckling/project/custom-attributes.component.ts]: This change is incorrect. You need to set this to something other than undefined.

File: src/duckling/project/project.service.ts [hunk 3]
```diff
@@ -136,13 +139,31 @@ export class ProjectService {
...
      */
-    addCustomAttribute(key : string, content : any) {
+    addCustomAttribute(key : string, content : JsonSchema) {
         let currentAttributes = this._project.customAttributes;
         let newAttribute = {key, content };
         let newAttributes = currentAttributes ? currentAttributes.concat([newAttribute]) : [].concat([newAttribute]);
         this._storeService.dispatch(changeCustomAttributes(newAttributes));
     }
 
+    hasCustomAttribute(attributeKey : AttributeKey) : boolean {
```
Comment (jeffswenson) [inline src/duckling/project/project.service.ts]: You should rename this isCustomAttribute. hasCustomAttribute sounds like a function that is checking to see if an entity has a custom attribute.

File: src/duckling/project/custom-attributes.component.ts (hunk 2 above)
Comment (tlein) [reply src/duckling/project/custom-attributes.component.ts]: fixed

File: src/duckling/project/project.service.ts (hunk 3 above)
Comment (tlein) [reply src/duckling/project/project.service.ts]: fixed

**Pull request merged**